lib/go/jsonframer/jsonframer.go

```diff
@@ -151,35 +153,77 @@ func ToFrame(jsonString string, options FramerOptions) (frame *data.Frame, err e
 	return getFrameFromResponseString(outString, options)
 }
 
-func GetRootData(jsonString string, rootSelector string) (string, error) {
-	if rootSelector != "" {
-		r := gjson.Get(string(jsonString), rootSelector)
-		if r.Exists() {
-			return r.String(), nil
-		}
-		expr, err := jsonata.Compile(rootSelector)
+func GetRootData(jsonString string, rootSelector string, framerType FramerType) (string, error) {
+	if rootSelector == "" {
+		return jsonString, nil
+	}
+	if framerType == FramerTypeJQ {
+		query, err := gojq.Parse(rootSelector)
 		if err != nil {
-			return "", errors.Join(ErrInvalidRootSelector, err)
-		}
-		if expr == nil {
-			return "", errors.Join(ErrInvalidRootSelector)
+			return "", errors.Join(ErrInvalidJQSelector, err)
 		}
 		var data any
 		err = json.Unmarshal([]byte(jsonString), &data)
 		if err != nil {
-			return "", errors.Join(ErrInvalidJSONContent, err)
+			return "", errors.Join(ErrUnMarshalingJSON, err)
 		}
-		res, err := expr.Eval(data)
-		if err != nil {
-			return "", errors.Join(ErrEvaluatingJSONata, err)
+		iter := query.Run(data)
+		out := []any{}
+		for {
+			v, ok := iter.Next()
+			if !ok {
+				break
+			}
+			if err, ok := v.(error); ok {
+				if err, ok := err.(*gojq.HaltError); ok && err.Value() == nil {
+					break
+				}
+				return "", errors.Join(ErrExecutingJQ, err)
+			}
+			out = append(out, v)
+		}
+		// if the result is array of 1 element array, ignore the outer array and return that single element from inner array
+		if len(out) == 1 {
+			if v, ok := out[0].([]any); ok {
+				outStr, err := json.Marshal(v)
+				if err != nil {
+					return "", errors.Join(ErrMarshalingJSON, err)
+				}
+				return string(outStr), nil
+			}
 		}
```

Comment:
@yesoreyeram, I’m wondering if this is actually beneficial. What if the length of the array changes based on the query and the user applies transformations? This could potentially cause issues, as the type of the value may change depending on the array’s contents.

@itsmylife